### ISSUE TITLE
WriteTag and IllegalBlockTag: crash on enum constant without Javadoc and missed Javadoc when modifier precedes annotation

### ISSUE THREAD

**Issue by codeAnqiang-ma** [checkstyle/checkstyle#21200]
I have read check documentation: https://checkstyle.org/checks/javadoc/writetag.html and https://checkstyle.org/checks/javadoc/illegalblocktag.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

Both `WriteTagCheck` and `IllegalBlockTagCheck` contain a private helper `findJavadoc` that locates the Javadoc comment of a declaration by probing a few fixed AST positions. Its assumptions about where a Javadoc block comment is attached in the AST are incomplete, which produces one crash and two silent false negatives. All three cases below share the same root cause.

### Case 1: WriteTag crashes with AssertionError on an enum constant without Javadoc

```bash
/var/tmp $ javac InputReproEnum.java
# successful compilation, no output

/var/tmp $ cat config-writetag-enum.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="WriteTag">
      <property name="tag" value="@since"/>
      <property name="tokens" value="ENUM_CONSTANT_DEF"/>
    </module>
  </module>
</module>

/var/tmp $ cat InputReproEnum.java
public enum InputReproEnum {

    /**
     * doc for FIRST
     */
    FIRST,

    SECOND,

    /**
     * doc for THIRD
     */
    @Deprecated
    THIRD
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-13.10.0-all.jar -c config-writetag-enum.xml InputReproEnum.java
Starting audit...
Exception in thread "main" java.lang.Error: Error was thrown while processing InputReproEnum.java
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:321)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:227)
	at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:429)
	at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:347)
	at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:205)
	at com.puppycrawl.tools.checkstyle.Main.main(Main.java:129)
Caused by: java.lang.AssertionError: Misuse of notNull: argument was null
	at com.puppycrawl.tools.checkstyle.utils.NullUtil.notNull(NullUtil.java:54)
	at com.puppycrawl.tools.checkstyle.checks.javadoc.WriteTagCheck.findJavadoc(WriteTagCheck.java:278)
	at com.puppycrawl.tools.checkstyle.checks.javadoc.WriteTagCheck.visitToken(WriteTagCheck.java:191)
```

`ENUM_CONSTANT_DEF` is listed as an acceptable token of `WriteTag` (both in the check documentation and in `WriteTagCheck.getAcceptableTokens()`), yet any enum constant that has no Javadoc kills the whole run: `findJavadoc` calls `NullUtil.notNull(ast.findFirstToken(TokenTypes.MODIFIERS))`, and `ENUM_CONSTANT_DEF` has an `ANNOTATIONS` child but no `MODIFIERS` child.

### Case 2: IllegalBlockTag silently skips Javadoc when a modifier precedes an annotation

```bash
/var/tmp $ javac InputRepro.java
# successful compilation, no output

/var/tmp $ cat config-illegalblocktag.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="IllegalBlockTag">
      <property name="tag" value="@todo"/>
    </module>
  </module>
</module>

/var/tmp $ cat InputRepro.java
public class InputRepro {

    /**
     * some doc
     * @todo remove this control case
     */
    @Deprecated
    public void annotationFirst() {
    }

    /**
     * some doc
     * @todo remove this bug case
     */
    public @Deprecated void modifierFirst() {
    }
}

/var/tmp $ java $RUN_LOCALE -jar checkstyle-13.10.0-all.jar -c config-illegalblocktag.xml InputRepro.java
Starting audit...
[ERROR] /var/tmp/InputRepro.java:5:8: Block tag 'todo' is matched illegal pattern [IllegalBlockTag]
Audit done.
Checkstyle ends with 1 errors.
```

Both methods carry the identical Javadoc with `@todo`; the only difference is `@Deprecated public` vs `public @Deprecated` (both are legal Java, and the latter ordering is common with type-use style annotations such as `public @Nullable String m()`). Only line 5 is reported; the `@todo` on line 13 is silently ignored. Expected: 2 violations.

The same false negative also happens for annotated enum constants (`/** @todo */ @Deprecated CONSTANT` is skipped while `/** @todo */ CONSTANT` is reported).

### Case 3: same false negative in WriteTag

```bash
/var/tmp $ cat config-writetag.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="WriteTag">
      <property name="tag" value="@since"/>
      <property name="tokens" value="METHOD_DEF"/>
    </module>
  </module>
</module>

/var/tmp $ cat InputReproWriteTag.java
public class InputReproWriteTag {

    /**
     * some doc without the required tag - control case
     */
    @Deprecated
    public void annotationFirst() {
    }

    /**
     * some doc without the required tag - bug case
     */
    public @Deprecated void modifierFirst() {
    }
}

/var/tmp $ java $RUN_LOCALE -jar checkstyle-13.10.0-all.jar -c config-writetag.xml InputReproWriteTag.java
Starting audit...
[ERROR] /var/tmp/InputReproWriteTag.java:6: Javadoc comment is missing @since tag. [WriteTag]
Audit done.
Checkstyle ends with 1 errors.
```

Expected: both methods reported as missing `@since`; actual: `modifierFirst()` is silently skipped.

---

**Describe what you expect in detail.**

1. `WriteTag` must not crash on valid Java code with a documented acceptable token. Enum constants without Javadoc should simply be skipped (case 1).
2. Both checks should find the Javadoc of a declaration regardless of the modifier/annotation order and regardless of the declaration kind, so cases 2 and 3 should each report 2 violations.

Root cause, verified against `master` (6fbc1bf): the duplicated helper `findJavadoc` (`WriteTagCheck.java:247-263`, `IllegalBlockTagCheck.java:216-234`) probes only 4 fixed attachment points and, when the `MODIFIERS` subtree contains an annotation, it unconditionally overwrites the comment already found at the `MODIFIERS` level with the lookup inside the first `ANNOTATION`:

```java
cmt = modifiers.findFirstToken(TokenTypes.BLOCK_COMMENT_BEGIN);
if (annotation != null) {
    cmt = annotation.findFirstToken(TokenTypes.BLOCK_COMMENT_BEGIN); // overwrites even when null
}
```

`checkstyle -T` shows why this matters: for `@Deprecated public void m()` the Javadoc is attached inside the `ANNOTATION` subtree, but for `public @Deprecated void m()` it is attached as a direct child of `MODIFIERS` (before `LITERAL_PUBLIC`), so the second lookup resets `cmt` to `null`. For annotated enum constants the comment lives under `ANNOTATIONS`/`ANNOTATION`, which is not probed at all, and the `WriteTag` variant additionally assumes every token has a `MODIFIERS` child, which does not hold for `ENUM_CONSTANT_DEF` (case 1 crash).

The project already has a battle-tested utility for exactly this job: `JavadocUtil.getAttachedJavadocComment(ast)` (used by `JavadocTypeCheck`, `JavadocMethodCheck`, `MissingJavadocMethodCheck`, `JavadocVariableCheck`). It recursively searches all children before the `IDENT` of the declaration, makes no `MODIFIERS` assumption, and filters with `isJavadocComment`. Replacing the duplicated `findJavadoc` with this utility fixes all three cases; I have this change with regression tests ready and can submit a PR once this issue is approved.

For context: before the migration to the Javadoc AST (#19150, merged 2026-06-25) `WriteTag` located Javadoc line-based via `FileContents.getJavadocBefore()` and handled all cases above correctly, so this is a regression of the migration that was then copied into the new `IllegalBlockTag` check (#21117). Possibly related to the already approved #20555 (another `findJavadoc` blind spot, compact source files), though the scenarios do not overlap.

This report was prepared with AI assistance; every reproduction and conclusion was verified locally by the author.

**Comment by varshini03** [comment]
Hi, I'd like to work on this issue!

**Comment by romani** [comment]
Crash should be fixed, 
Fix is welcome

**Comment by romani** [comment]
Fixed by https://github.com/checkstyle/checkstyle/pull/21234
